R/rdeephaven/src/client.cpp

```diff
@@ -438,32 +455,34 @@ class ClientWrapper {
         auto ticket = internal_tbl_hdl_mngr.NewTicket();
         auto fd = deephaven::client::utility::ConvertTicketToFlightDescriptor(ticket);
 
-        deephaven::client::utility::OkOrThrow(DEEPHAVEN_EXPR_MSG(wrapper.FlightClient()->DoPut(options, fd, schema, &fsw, &fmr)));
+        deephaven::client::utility::OkOrThrow(DEEPHAVEN_LOCATION_EXPR(wrapper.FlightClient()->DoPut(options, fd, schema, &fsw, &fmr)));
         while(true) {
             std::shared_ptr<arrow::RecordBatch> this_batch;
-            deephaven::client::utility::OkOrThrow(DEEPHAVEN_EXPR_MSG(record_batch_reader->ReadNext(&this_batch)));
+            deephaven::client::utility::OkOrThrow(DEEPHAVEN_LOCATION_EXPR(record_batch_reader->ReadNext(&this_batch)));
             if (this_batch == nullptr) {
                 break;
             }
-            deephaven::client::utility::OkOrThrow(DEEPHAVEN_EXPR_MSG(fsw->WriteRecordBatch(*this_batch)));
+            deephaven::client::utility::OkOrThrow(DEEPHAVEN_LOCATION_EXPR(fsw->WriteRecordBatch(*this_batch)));
         }
-        deephaven::client::utility::OkOrThrow(DEEPHAVEN_EXPR_MSG(fsw->DoneWriting()));
-        deephaven::client::utility::OkOrThrow(DEEPHAVEN_EXPR_MSG(fsw->Close()));
+        deephaven::client::utility::OkOrThrow(DEEPHAVEN_LOCATION_EXPR(fsw->DoneWriting()));
+        deephaven::client::utility::OkOrThrow(DEEPHAVEN_LOCATION_EXPR(fsw->Close()));
 
         auto new_tbl_hdl = internal_tbl_hdl_mngr.MakeTableHandleFromTicket(ticket);
         return new TableHandleWrapper(new_tbl_hdl);
     }
 
     void Close() {
-        internal_client.Close();
+        internal_client->Close();
     }
 
 private:
-    deephaven::client::Client internal_client;
-    const deephaven::client::TableHandleManager internal_tbl_hdl_mngr = internal_client.GetManager();
+    // We let R manage the lifetime of internal_client underlying C++ object,
+    // according to its tracking of references.
+    // We hold one here, but there may be other references in the case of the enterprise client.
+    Rcpp::XPtr<deephaven::client::Client> internal_client;
+    const deephaven::client::TableHandleManager internal_tbl_hdl_mngr = internal_client->GetManager();
```

Comment:
Because now it is a pointer?   So otherwise it won't compile?  :-)
More strictly speaking, is a pointer-like object, which in terms of behavior means it implements both the `->` and `*` operators.  So essentially `XPtr` in R behaves like, eg, `shared_ptr` in standard C++.  It behaves like a pointer to an object (the contained object).
The difference is `XPtr` uses the internal R engine mechanism for tracking object lifetimes, and depending on how you crate it, it will delete the underlying object when R does no longer reference them.   The default construction does delete.